lib/Sema/TypeCheckAttr.cpp

```diff
@@ -1217,17 +1217,49 @@ visitDynamicMemberLookupAttr(DynamicMemberLookupAttr *attr) {
   // This attribute is only allowed on nominal types.
   auto decl = cast<NominalTypeDecl>(D);
   auto type = decl->getDeclaredType();
-  
+  auto &ctx = decl->getASTContext();
+
   // Look up `subscript(dynamicMember:)` candidates.
   auto subscriptName = DeclName(TC.Context, DeclBaseName::createSubscript(),
                                 TC.Context.Id_dynamicMember);
   auto candidates = TC.lookupMember(decl, type, subscriptName);
-  
-  // If there are no candidates, then the attribute is invalid.
+
+  // If there are no candidates, then it is possible that the subscript was
+  // defined without an argument label. For example:
+  //
+  // ```
+  // subscript(dynamicMember: KeyPath<A, B>) -> C
+  // ```
+  //
+  // So, let's lookup again, but this time explicitly check for this case.
   if (candidates.empty()) {
+    auto subscriptParam = new (ctx) ParamDecl(
+        ParamDecl::Specifier::Default, SourceLoc(), SourceLoc(), Identifier(),
+        SourceLoc(), TC.Context.Id_dynamicMember, decl->getDeclContext());
+    auto subscriptParamList =
+        ParameterList::create(ctx, SourceLoc(), {subscriptParam}, SourceLoc());
+    auto explicitSubscriptName = DeclName(
+        TC.Context, DeclBaseName::createSubscript(), subscriptParamList);
+    auto newCandidates = TC.lookupMember(decl, type, explicitSubscriptName);
```

Comment:
I'd suggest instead of doing all this to do a member lookup using `DeclBaseName::createSubscript()` and use `isValidDynamicMemberLookupSubscript` with a slight modification to ignore labels while checking validity in this case.

Reply:
That's a nice idea. I think we can just lookup all subscripts, but instead of looking at the argument name, we can look at the parameter name while validating the subscript.

Reply:
Oh, I now realise what you meant. Let me make the changes.

Reply:
Ok done.